runtime/ops/something.cc

```diff
@@ -13,44 +15,207 @@ namespace runtime {
+    const double in_shape_h = static_cast<double>(AsScalar(in_shape.At({2})));
+    const double in_shape_w = static_cast<double>(AsScalar(in_shape.At({3})));
+    std::vector<chainerx::Array> rois_list;
+    std::vector<chainerx::Array> roi_indices_list;
+    const int64_t batch_size = static_cast<int64_t>(chainerx::AsScalar(in_shape.At({0})));
+    for (int64_t b = 0; b < batch_size; ++b) {
+        std::vector<chainerx::Array> rois_list_per_batch;
+        std::vector<double> confs_list_per_batch;
+        for (size_t l = 0; l < scales.size(); ++l) {
+            const float* loc_l = static_cast<float*>(locs[l].raw_data()) + b * locs[l].shape()[1] * locs[l].shape()[2];
+            const float* conf_l = static_cast<float*>(confs[l].raw_data()) + b * confs[l].shape()[1];
+            const auto& h_l = hs[l];
+            const int64_t k_l = locs[l].shape()[1];
+            std::vector<chainerx::Array> roi_l_list;
+            for (size_t u = 0; u < h_l.shape()[2]; ++u) {
+                for (size_t v = 0; v < h_l.shape()[3]; ++v) {
+                    for (size_t iar = 0; iar < k_anchor_ratios.size(); ++iar) {
+                        double ar = k_anchor_ratios[iar];
+                        const double ay = (u + 0.5) / scales[l];
+                        const double ax = (v + 0.5) / scales[l];
+                        const double w = std::round(1.0 / scales[l] / std::sqrt(ar));
+                        const double h = std::round(w * ar);
+                        const double ah = h * (k_anchor_size << l) * scales[l];
+                        const double aw = w * (k_anchor_size << l) * scales[l];
+
+                        const int64_t k = h_l.shape()[3] * k_anchor_ratios.size() * u + k_anchor_ratios.size() * v + iar;
+                        const double loc_l_y = *(loc_l + k * 4 + 0);
+                        const double loc_l_x = *(loc_l + k * 4 + 1);
+                        const double loc_l_h = *(loc_l + k * 4 + 2);
+                        const double loc_l_w = *(loc_l + k * 4 + 3);
+
+                        const double roi_l_y = ay + ah * loc_l_y;
+                        const double roi_l_x = ax + aw * loc_l_x;
+                        const double roi_l_h = ah * std::exp(std::min(loc_l_h, k_exp_clip));
+                        const double roi_l_w = aw * std::exp(std::min(loc_l_w, k_exp_clip));
 
-    WARN_ONCE("TODO: Implement ChainerCVRPNDecode");
+                        // yxhw -> tlbr (top left, bottom right)
+                        const double roi_l_tly = std::max(roi_l_y - roi_l_h / 2.0, 0.0);
+                        const double roi_l_tlx = std::max(roi_l_x - roi_l_w / 2.0, 0.0);
+                        const double roi_l_bry = std::min(roi_l_y + roi_l_h / 2.0, in_shape_h);
+                        const double roi_l_brx = std::min(roi_l_x + roi_l_w / 2.0, in_shape_w);
+                        auto data = std::shared_ptr<void>(new float[4]{static_cast<float>(roi_l_tly),
```

Comment:
I'm not sure, but cannot we avoid the redundant heap allocation by argsorting confidences first?

Reply:
Yes.